evennia/contrib/slots_handler/slots.py

```diff
@@ -0,0 +1,455 @@
+    remove the target from any of the slots given it. Both of these methods
+    have the option to be given a custom `slots` argument, which will override
+    the slots on the held object.
+
+    Using the custom `slots` in `attach()` and `drop()` provides some
+    customization facility in that you can store *any* object in a slot, not
+    just an object that has been set up for it.
+    """
+
+    def __init__(self, obj):
+        self.obj = obj
+        self._objid = obj.id
+
+    def __defrag_nums(self, cats):
+        "Worker function to consolidate filled numbered slots."
+        if not isinstance(cats, list):
+            cats = [cats]
+
+        arrays = self.all(obj=True)
+        arrays = [a for a in arrays if a.key in cats]
+
+        for a in arrays:
+            slots = a.value
+            out = {k: v for k, v in slots.items() if isinstance(k, str)}
+            numbered = [(k, v) for k, v in slots.items() if isinstance(k, int)]
+            keys = [n[0] for n in numbered]
+            values = [n[1] for n in numbered]
+            d = len(values) - 1
+            empty = [values.pop(0) for i in range(0, d) if values[0] == ""]
+            numbered = zip(keys, values + empty)
+            out.update(numbered)
+            self.obj.attributes.add(a.key, out, category="slots")
+
+    # Public methods
+    def all(self, obj=False):
+        """
+        Return all slots.
+
+        Args:
+            obj (bool): Whether or not to return the attribute objects.
+                (Default: False)
+
+        Returns:
+            slots (dict): A dict of all slots.
+        """
+        d = self.obj.attributes.get(category="slots", return_obj=True)
+
+        if not d:
+            return {}
+        elif not isinstance(d, list):
+            d = [d]
+
+        if obj:
+            # Return attribute objects if requested.
+            return d
+        else:
+            # Return a dict detached from the database.
+            r = {s.key: s.value for s in d}
+            return r
+
+    def add(self, slots):
+        """
+        Create arrays of slots, or add additional slots to existing arrays.
+
+        Args:
+            slots (dict): A dict of slots to add. Since you can't add empty
+                categories, it would be pointless to pass a list to this
+                function, and so it doesn't accept lists for input.
+
+        Returns:
+            slots (dict): A dict of slots that have been successfully added.
+        """
+
+        if not isinstance(slots, (dict, _SaverDict)):
+            raise Exception("You have to declare slots in the form "
+                            "`{key: [values]}`.")
+
+        wiz.msg("Gonna add some slots! Yeah!")
+
+        arrays = {a.key: a for a in self.obj.slots.all(obj=True)}
+        modified = {}
+        for name in slots:
+            existing = arrays.get(name, False)
+            wiz.msg("-> Inspecting {}: {}".format(name, existing))
```

Comment:
Apart from `wiz` not being defined, these look like a debug statements which should probably be removed for the finished contrib?